File: Problemset/0654-Maximum Binary Tree/Recursion.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        if nums:
            maxnum = max(nums)
            root = TreeNode(maxnum)
            index = nums.index(maxnum)
            root.left = self.constructMaximumBinaryTree(nums[:index])
            root.right = self.constructMaximumBinaryTree(nums[index + 1:])
            return root
        else:
            return None

File: Problemset/0654-Maximum Binary Tree/Recursion.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {Integer[]} nums
# @return {TreeNode}
def construct_maximum_binary_tree(nums)
    if not nums.empty?
        maxnum = nums.max
        root = TreeNode.new(maxnum)
        index = nums.index(maxnum)
        root.left = construct_maximum_binary_tree(nums[0...index])
        root.right = construct_maximum_binary_tree(nums[index + 1..-1])
        return root
    else
        return nil
    end
end
